Document/CategoryManager.php

```diff
@@ -11,14 +11,125 @@
 
 namespace Sonata\ClassificationBundle\Document;
 
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Sonata\ClassificationBundle\Model\CategoryInterface;
 use Sonata\ClassificationBundle\Model\CategoryManagerInterface;
+use Sonata\ClassificationBundle\Model\ContextInterface;
+use Sonata\ClassificationBundle\Model\ContextManagerInterface;
 use Sonata\CoreBundle\Model\BaseDocumentManager;
 use Sonata\DatagridBundle\Pager\Doctrine\Pager;
 use Sonata\DatagridBundle\ProxyQuery\Doctrine\ProxyQuery;
 
 class CategoryManager extends BaseDocumentManager implements CategoryManagerInterface
 {
     /**
+     * @var CategoryInterface[]
+     */
+    protected $categories;
+
+    /**
+     * @var ContextManagerInterface
+     */
+    protected $contextManager;
+
+    /**
+     * @param string                  $class
+     * @param ManagerRegistry         $registry
+     * @param ContextManagerInterface $contextManager
+     */
+    public function __construct($class, ManagerRegistry $registry, ContextManagerInterface $contextManager)
+    {
+        parent::__construct($class, $registry);
+
+        $this->contextManager = $contextManager;
+        $this->categories = array();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRootCategoriesPager($page = 1, $limit = 25, $criteria = array())
+    {
+        $queryBuilder = $this->getObjectManager()->createQueryBuilder()
+            ->select('c')
+            ->from($this->class, 'c')
+            ->andWhere('c.parent IS NULL');
+
+        $pager = new Pager($limit);
+        $pager->setQuery(new ProxyQuery($queryBuilder));
+        $pager->setPage($page);
+        $pager->init();
+
+        return $pager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSubCategoriesPager($categoryId, $page = 1, $limit = 25, $criteria = array())
+    {
+        $queryBuilder = $this->getObjectManager()->createQueryBuilder()
+            ->select('c')
+            ->from($this->class, 'c')
+            ->where('c.parent = :categoryId')
+            ->setParameter('categoryId', $categoryId);
+
+        $pager = new Pager($limit);
+        $pager->setQuery(new ProxyQuery($queryBuilder));
+        $pager->setPage($page);
+        $pager->init();
+
+        return $pager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRootCategory($context = null)
+    {
+        $context = $this->getContext($context);
+
+        $this->loadCategories($context);
+
+        return $this->categories[$context->getId()][0];
```

Comment:
Use current() is what you want is the first element. You never now if `$categories` passes through a call to `array_filter()`, for some reason…